src/algorithm/SCAN-Cuda.cpp

```diff
@@ -0,0 +1,122 @@
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+// Copyright (c) 2017-21, Lawrence Livermore National Security, LLC
+// and RAJA Performance Suite project contributors.
+// See the RAJAPerf/COPYRIGHT file for details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+
+#include "SCAN.hpp"
+
+#include "RAJA/RAJA.hpp"
+
+#if defined(RAJA_ENABLE_CUDA)
+
+#include "cub/device/device_scan.cuh"
+#include "cub/util_allocator.cuh"
+
+#include "common/CudaDataUtils.hpp"
+
+#include <iostream>
+
+namespace rajaperf
+{
+namespace algorithm
+{
+
+  //
+  // Define thread block size for CUDA execution
+  //
+  const size_t block_size = 256;
+
+
+#define SCAN_DATA_SETUP_CUDA \
+  allocAndInitCudaDeviceData(x, m_x, iend); \
+  allocAndInitCudaDeviceData(y, m_y, iend);
+
+#define SCAN_DATA_TEARDOWN_CUDA \
+  getCudaDeviceData(m_y, y, iend); \
+  deallocCudaDeviceData(x); \
+  deallocCudaDeviceData(y);
+
+
+void SCAN::runCudaVariant(VariantID vid)
+{
+  const Index_type run_reps = getRunReps();
+  const Index_type ibegin = 0;
+  const Index_type iend = getRunSize();
+
+  SCAN_DATA_SETUP;
+
+  if ( vid == Base_CUDA ) {
+
+    SCAN_DATA_SETUP_CUDA;
+
+    cudaStream_t stream = 0;
+
+    RAJA::operators::plus<Real_type> binary_op;
+    Real_type init_val = 0.0;
+
+    int len = iend - ibegin;
+
+    // Determine temporary device storage requirements
+    void* d_temp_storage = nullptr;
+    size_t temp_storage_bytes = 0;
+    cudaErrchk(::cub::DeviceScan::ExclusiveScan(d_temp_storage,
+                                                temp_storage_bytes,
+                                                x+ibegin,
+                                                y+ibegin,
+                                                binary_op,
+                                                init_val,
+                                                len,
+                                                stream));
+
+    // Allocate temporary storage
+    unsigned char* temp_storage;
+    allocCudaDeviceData(temp_storage, temp_storage_bytes);
+    d_temp_storage = temp_storage;
+
+    startTimer();
```

Comment:
should the startTimer call go up to line 62 for a fair comparison with the RAJA variant?

Reply:
I agree this is a bit asymmetric, but I didn't want to time the cudaMalloc. The RAJA variant gets memory from a pool that already has memory, but this variant calls cuda Malloc. The RAJA variant queries how much memory it needs each rep, but this variant only queries how much memory it needs once outside the timer. So this variant does do less work each rep than the RAJA variant which is not ideal, but I wasn't sure how to avoid timing the cudaMalloc and get the memory query into the rep loop.

Reply:
OK. I can live with that.